plugins/BEdita/Core/src/Model/Table/FoldersTable.php

```diff
@@ -106,6 +115,40 @@ public function hasAtMostOneParent(Folder $entity)
         return $rule($entity, ['operator' => '==', 'count' => 1]) && !empty($entity->parent->id);
     }
 
+    /**
+     * Custom rule to check if the folder entity is restorable
+     * i.e. if its parents have not been deleted.
+     *
+     * If entity is new or `deleted` is not dirty (no change) or it is equal to true (delete action) then return true.
+     *
+     * @param Folder $entity The entity to check
+     * @return bool
+     */
+    public function isFolderRestorable(Folder $entity)
+    {
+        if ($entity->isNew() || !$entity->isDirty('deleted') || $entity->deleted === true) {
+            return true;
+        }
+
+        $node = $this->TreeNodes
+            ->find()
+            ->where([$this->TreeNodes->aliasField('object_id') => $entity->id])
+            ->firstOrFail();
+
+        $deletedParents = $this->find()
+            ->innerJoinWith('TreeNodes', function (Query $query) use ($node) {
+                return $query->where(function (QueryExpression $exp) use ($node) {
+                    return $exp
+                        ->lt($this->TreeNodes->aliasField('tree_left'), $node->get('tree_left'))
+                        ->gt($this->TreeNodes->aliasField('tree_right'), $node->get('tree_right'));
+                });
+            })
+            ->where([$this->aliasField('deleted') => true])
+            ->count();
```

Comment:
Can this be rewritten using `Table::exists()`? It shouldn't be a problem on trees with a limited depth, though.

Reply:
I'm not sure I can replicate that query using `exists()` method because it accept only `$conditions` argument used from query builder in `where()` method.

I can give it a try or replicate what `exists()` do i.e. count the result of
```
->select(['existing' => 1])
->limit(1)
 ->enableHydration(false)
 ->toArray()
```

Reply:
Uhm… I'm not sure anymore. This query is only executed in very rare cases, and using `count()` vs `exists()` doesn't add a significant overhead. I'd go with the most readable option.